### PR TITLE
FEATURE: Replace frontend build stack with webpack

### PR DESCRIPTION
**Pull request by jonnitto**
Resolves: #69

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -1,5 +1,5 @@
 {
-    "name": "neos/demo",
+    "name": "neos/test",
```
Comment (mficzel) [inline composer.json]: Why is this changed?

File: .nvmrc [hunk 2]
```diff
@@ -0,0 +1 @@
+11
```
Comment (mficzel) [inline .nvmrc]: Why this, i thought the current best practice is requiring the node version on the package.json.

Comment (Sebobo) [reply .nvmrc]: This doesn't seem to be supported by nvm yet: https://github.com/nvm-sh/nvm/issues/651

File: .stylelintrc [hunk 3]
```diff
@@ -0,0 +1,48 @@
+{
+  "ignoreFiles": [
+    "**/Public/**",
+    "**/Private/Templates/**",
+    "node_modules/**",
+    "**/*.noLinter.*"
+],
+  "extends": "stylelint-config-standard",
```
Comment (Sebobo) [inline .stylelintrc]: Wouldn't `stylelint-config-recommended-scss` be enough?
I'm scared that these big config files might intimidate people or confuse them.

File: .yarnclean [hunk 4]
```diff
@@ -0,0 +1,44 @@
+# Exclude some needed stuff
```
Comment (Sebobo) [inline .yarnclean]: Do we need this file with all it's unrelated entries?

File: Resources/Private/Fusion/Document/Base.fusion [hunk 5]
```diff
@@ -11,13 +11,13 @@ prototype(Neos.Demo:Document.Base) < prototype(Neos.Fusion:Component) {
...
             }
+            javascripts {
+                site = Neos.Demo:Document.Fragment.JavaScripts
+            }
             metadata = Neos.Demo:Document.Fragment.MetaData
         }
 
         bodyTag.attributes.class = ${props.bodyClass}
         body = ${props.body}
-
```
Comment (Sebobo) [inline Resources/Private/Fusion/Document/Base.fusion]: Why are they moved to the head?

File: Resources/Private/Fusion/Main.scss [hunk 6]
```diff
@@ -0,0 +1,18 @@
+// Include libraries
```
Comment (Sebobo) [inline Resources/Private/Fusion/Main.scss]: Why is Main and Variables not in the Styles folder?

File: Resources/Private/JavaScript/main.js [hunk 7]
```diff
@@ -1,29 +1,11 @@
...
-})(window, jQuery);
+(function($) {
+    $('.top-navigation-wrap').addClass('navbar-fixed-top');
+
+    /**
+     * Check if page is scrolled and switch 'scrolling' class for top navigation.
+     * In our case this adds a nice shadow effect.
+     */
+    $(window).on('scrollstart scrollend', function() {
+        $topNavigationWrap.toggleClass('scrolling', $(document).scrollTop() > 0);
```
Comment (Sebobo) [inline Resources/Private/JavaScript/main.js]: After this change `$topNavigationWrap` is not defined anymore?

File: package.json [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    "postcss-responsive-type": "^1.0.0",
+    "postcss-round-subpixels": "^1.2.0",
+    "postcss-url": "^8.0.0",
+    "postcss-vmax": "^1.0.0",
+    "prettier": "^1.18.2",
+    "prettier-loader": "^3.2.0",
+    "sass-loader": "^8.0.0",
+    "sort-css-media-queries": "^1.4.3",
+    "terser-webpack-plugin": "^2.1.3",
+    "ts-loader": "^6.2.0",
```
Comment (Sebobo) [inline package.json]: If we don't use typescript and other things in this package, we also shouldn't have all those dependencies

File: webpack.packages.ts [hunk 9]
```diff
@@ -0,0 +1,13 @@
+/**
+ * Pass an object with the following properties to the array:
+ *
+ * packageName   <string>  (optional) If you are not in a Monorepo (and in a package), leave this value blank
+ * filename      <string>  (optional) (default: `Main.js`) The name of the entry file
+ * entryPath     <string>  (optional) (default: `Resources/Private/Fusion`) The entry path, relative to the package
+ * publicPath    <string>  (optional) (default: `Resources/Public`) The public path, relative to the package
+ * hasSourceMap  <boolean> (optional) (default: `true`) Flag to toggle source map generation
+ */
+
```
Comment (Sebobo) [inline webpack.packages.ts]: This one would need some documentation.

File: composer.json (hunk 1 above)
Comment (jonnitto) [reply composer.json]: Ups, my fault

File: .yarnclean (hunk 4 above)
Comment (jonnitto) [reply .yarnclean]: This file helps to make the `node_modules` folder a bit smaller after an install

File: Resources/Private/Fusion/Document/Base.fusion (hunk 5 above)
Comment (jonnitto) [reply Resources/Private/Fusion/Document/Base.fusion]: Because it is better to have them in the head with the attribute `defer`: Start loading and wait till the document is ready.

File: Resources/Private/Fusion/Main.scss (hunk 6 above)
Comment (jonnitto) [reply Resources/Private/Fusion/Main.scss]: Because it is the first step for the refactoring. The goal is the get ridden of the Styles folder, that's why it has the same entry point (`Private/Fusion`)

File: Resources/Private/JavaScript/main.js (hunk 7 above)
Comment (jonnitto) [reply Resources/Private/JavaScript/main.js]: Fixed, thanks

File: package.json (hunk 8 above)
Comment (jonnitto) [reply package.json]: I just want it to have the same as Neos.io. Currently it is used as the webpack config

File: composer.json (hunk 1 above)
Comment (jonnitto) [reply composer.json]: Fixed.

File: .nvmrc (hunk 2 above)
Comment (jonnitto) [reply .nvmrc]: Yep, currently, it is better the have the `.nvmrc` file in a repo

File: .stylelintrc (hunk 3 above)
Comment (jonnitto) [reply .stylelintrc]: Good point. I've updated the file. I left some rules inside, as they are for PostCSS rules

File: package.json (hunk 8 above)
Comment (jonnitto) [reply package.json]: I have removed TypeScript and all React

File: Resources/Private/Fusion/Main.scss (hunk 6 above)
Comment (Sebobo) [reply Resources/Private/Fusion/Main.scss]: But don't we still need a Styles folder for global stuff like variables, typography and such?

Comment (Sebobo) [reply Resources/Private/Fusion/Main.scss]: We never had files directly in Private

**Comment by Sebobo** [comment]
I think it would be great a have all the build system explained in the readme and where to put files.

**Comment by jonnitto** [comment]
> I think it would be great a have all the build system explained in the readme and where to put files.

Will do.

**Comment by davidspiola** [comment]
Hi Folks, is this someting to mention in the 5.1 release?

**Comment by Sebobo** [comment]
With a bit of CSS refreshment, this could be release notes worthy yes :)

**Comment by bwaidelich** [comment]
I added an issue for this in order to simplify release management and to collect higher-level details for the release notes: #69

File: .gitignore [hunk 10]
```diff
@@ -1,3 +1,5 @@
-.sass-cache/
+.sass-cache
+.vscode
+.idea
 
 /Resources/Private/node_modules
```
Comment (mficzel) [inline .gitignore]: this should be adjusted to  `node_modules` i assume

File: package.json [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    "> 0.5%",
+    "last 2 versions",
+    "Firefox ESR",
+    "not ie <= 10",
+    "not ie_mob <= 10",
+    "not bb <= 10",
+    "not op_mob <= 12.1"
+  ],
+  "scripts": {
+    "start": "webpack --watch",
```
Comment (mficzel) [inline package.json]: I would expect to call this `watch` but this may be my personal experiecne. I am fine if others call it differently, just give this name one more thought.

**Comment by mficzel** [comment]
I added the node_modules to the .gitignore. So i would be fine with merging this now. I do not care that much wether it is called `yarn start` or yarn `watch` and we can always change this.

File: Resources/Private/Fusion/Document/Fragment/StyleSheets.fusion [hunk 12]
```diff
@@ -3,8 +3,8 @@
  */
 prototype(Neos.Demo:Document.Fragment.StyleSheets) < prototype(Neos.Fusion:Component) {
     renderer = afx`
+        <link href="https://fonts.googleapis.com/css?family=Work+Sans:400&display=swap" rel="stylesheet" />
```
Comment (Sebobo) [inline Resources/Private/Fusion/Document/Fragment/StyleSheets.fusion]: Can we prevent having to load a font from Google?

Comment (jonnitto) [reply Resources/Private/Fusion/Document/Fragment/StyleSheets.fusion]: We could also include these files, yes. But in my opinion, we can make the package with this change a bit smaller…

Comment (jonnitto) [reply Resources/Private/Fusion/Document/Fragment/StyleSheets.fusion]: Done

**Comment by Sebobo** [comment]
Awesome! Works :)

With an updated readme I will approve ;)

**Comment by mficzel** [comment]
@Sebobo are you fine with the updated readme. Then this should be ready

**Comment by Sebobo** [comment]
Was just finished reading when somebody merged...

Awesome! Thx @jonnitto
